projects/opendr_ws_2/src/opendr_perception/opendr_perception/panoptic_segmentation_efficient_ps.py

```diff
@@ -0,0 +1,192 @@
+        config_file = Path(sys.modules[
+                               EfficientPsLearner.__module__].__file__).parent / 'configs' / 'singlegpu_cityscapes.py'
+        self._learner = EfficientPsLearner(str(config_file))
+
+        # Other
+        self._tmp_folder = Path(__file__).parent.parent / 'tmp' / 'efficientps'
+        self._tmp_folder.mkdir(exist_ok=True, parents=True)
+
+    def _init_learner(self) -> bool:
+        """
+        The model can be initialized via
+        1. downloading pre-trained weights for Cityscapes or KITTI.
+        2. passing a path to an existing checkpoint file.
+
+        This has not been done in the __init__() function since logging is available only once the node is registered.
+        """
+        if self.checkpoint in ['cityscapes', 'kitti']:
+            file_path = EfficientPsLearner.download(str(self._tmp_folder),
+                                                    trained_on=self.checkpoint)
+            self.checkpoint = file_path
+
+        if self._learner.load(self.checkpoint):
+            self.get_logger().info('Successfully loaded the checkpoint.')
+            return True
+        else:
+            self.get_logger().error('Failed to load the checkpoint.')
+            return False
+
+    def _init_subscriber(self):
+        """
+        Subscribe to all relevant topics.
+        """
+        self.image_subscriber = self.create_subscription(ROS_Image, self.input_rgb_image_topic,
+                                                         self.callback, 1)
+
+    def _init_publisher(self):
+        """
+        Set up the publishers as requested by the user.
+        """
+        if self.output_heatmap_topic is not None:
+            self._instance_heatmap_publisher = self.create_publisher(ROS_Image,
+                                                                     f'{self.output_heatmap_topic}/instance',
+                                                                     10)
+            self._semantic_heatmap_publisher = self.create_publisher(ROS_Image,
+                                                                     f'{self.output_heatmap_topic}/semantic',
+                                                                     10)
+        if self.output_rgb_visualization_topic is not None:
+            self._visualization_publisher = self.create_publisher(ROS_Image,
+                                                                  self.output_rgb_visualization_topic,
+                                                                  10)
+
+    def listen(self):
+        """
+        Start the node and begin processing input data. The order of the function calls ensures that the node does not
+        try to process input images without being in a trained state.
+        """
+        self.get_logger().info('EfficientPS node started!')
+        if self._init_learner():
+            self._init_publisher()
+            self._init_subscriber()
+            rclpy.spin(self)
```

Comment:
I would suggest moving the logging message after the initialization so in case there's an error in the 3 init methods, the message is not printed.

```suggestion
        if self._init_learner():
            self._init_publisher()
            self._init_subscriber()
            self.get_logger().info('EfficientPS node started!')
            rclpy.spin(self)
```